### PR TITLE
find un-managed mocks

### LINKED ISSUE #61: find un-managed mocks with `gomockhandler check` cmd
find mocks un-managed by gomockhandler.
IMO, this feature should be implemented in `gomockhandler check` to keep gomockhandler simple.

### PR DESCRIPTION
**Pull request by riita10069**
This PR ensures that the check command has a one-to-one correspondence between gomockhandler.json and the actual mock generated by gomock.
In other words, it finds zombie mocks. (A zombie mock is a mock generated by gomock but not managed by gomockhandler.)
And It can also find mocks which are managed by gomockhandler but not generated

---

Fixes #61

### CODE REVIEW

**Comment by riita10069** [comment]
#61

File: internal/command/check.go [hunk 1]
```diff
@@ -93,9 +111,14 @@ func (r Runner) Check() {
 	if err != nil {
 		log.Fatalf("failed to run: %v", err.Error())
 	}
+
+	if zombieList := zombieChecker.Check(); len(zombieList) != 0 {
+		fmt.Fprintf(os.Stderr, "There are un-managed mocks!!\nThe list of them: %v\n", zombieList)
```
Comment (sanposhiho) [inline internal/command/check.go:116]: nits

```suggestion
		fmt.Fprintf(os.Stderr, "[WARN] Some mocks are un-managed by gomockhandler: %v\n", zombieList)
```

File: internal/zombie/checker.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (c *Checker) setHash(ctx context.Context, trees []string) error {
+	eg, _  := errgroup.WithContext(ctx)
+	sem := semaphore.NewWeighted(ConcurrencyNum)
+
+	for _, path := range trees {
+		path := path
+		sem.Acquire(ctx, 1)
+		eg.Go(func() error {
```
Comment (sanposhiho) [inline internal/zombie/checker.go:71]: We have to use `sem.Release(1)`after this line otherwise deadlock

```
defer sem.Release(1)
```

File: internal/zombie/checker.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			lines, err := util.ReadTwoLines(path)
+			if err != nil {
+				return fmt.Errorf("failed to readline to find a mock: %w", err)
+			}
+			if lines[0] == CommentGeneratedByGoMock {
+				c.mux.Lock()
+				key := lines[1][11:]
```
Comment (sanposhiho) [inline internal/zombie/checker.go:78]: It is better to use the path of the mock file as a key. 
This is because we cannot get this key when mockgen change the format of mock file in the future. And we don't want to limit the versions of gomock that users can use.

File: internal/command/check.go [hunk 4]
```diff
@@ -30,17 +31,34 @@ func (r Runner) Check() {
...
 	for _, m := range ch.Mocks {
 		m := m
 
 		var runner mockgen.Runner
 		g.Go(func() error {
 			switch m.Mode {
 			case model.ReflectMode:
+				key := m.ReflectModeRunner.PackageName + " (interfaces: " + m.ReflectModeRunner.Interfaces + ")"
+				if ok := zombieChecker.Search(key); !ok {
+					fmt.Fprintf(os.Stderr, "[ERROR] mock is not up to date. destination: %s\n", key)
```
Comment (sanposhiho) [inline internal/command/check.go:50]: It is good idea, we can find configured but un-generated mock. 👍 

nits:

```suggestion
					fmt.Fprintf(os.Stderr, "[ERROR] mock has not been generated. destination: %s\n", key)
```

File: internal/command/check.go [hunk 5]
```diff
@@ -30,17 +31,34 @@ func (r Runner) Check() {
...
+					isFail = true
+				}
+
 				runner = m.ReflectModeRunner
 			case model.SourceMode:
+				key := m.SourceModeRunner.Source
+				if ok := zombieChecker.Search(key); !ok {
+					fmt.Fprintf(os.Stderr, "[ERROR] mock is not up to date. destination: %s\n", key)
```
Comment (sanposhiho) [inline internal/command/check.go:58]: nits:

```suggestion
					fmt.Fprintf(os.Stderr, "[ERROR] mock has not been generated. destination: %s\n", key)
```

**Review by sanposhiho — CHANGES_REQUESTED** [review]
Thanks for nice contribution! 

I left some comments. Please check when you are available 👍

File: internal/command/check.go [hunk 6]
```diff
@@ -30,17 +31,34 @@ func (r Runner) Check() {
...
+	g, _ := errgroup.WithContext(ctx)
 	for _, m := range ch.Mocks {
 		m := m
 
 		var runner mockgen.Runner
 		g.Go(func() error {
 			switch m.Mode {
 			case model.ReflectMode:
+				key := m.ReflectModeRunner.PackageName + " (interfaces: " + m.ReflectModeRunner.Interfaces + ")"
+				if ok := zombieChecker.Search(key); !ok {
```
Comment (sanposhiho) [inline internal/command/check.go:49]: We have to do this check after this line: https://github.com/sanposhiho/gomockhandler/blob/master/internal/command/check.go#L71

This is because gomockhandler has the option `target_dir`.

File: internal/command/check.go [hunk 7]
```diff
@@ -30,17 +31,34 @@ func (r Runner) Check() {
...
+					fmt.Fprintf(os.Stderr, "[ERROR] mock is not up to date. destination: %s\n", key)
+					isFail = true
+				}
+
 				runner = m.ReflectModeRunner
 			case model.SourceMode:
+				key := m.SourceModeRunner.Source
+				if ok := zombieChecker.Search(key); !ok {
```
Comment (sanposhiho) [inline internal/command/check.go:57]: (same as https://github.com/sanposhiho/gomockhandler/pull/65/files#r711660619)

**Comment by sanposhiho** [comment]
And you may already know about it, but we can check the behaviour of gomockhandler on /playground dir.

I wrote a contribution guide for that.
https://github.com/sanposhiho/gomockhandler/blob/master/CONTRIBUTING.md

File: internal/zombie/checker.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func NewChecker() *Checker {
+	return &Checker{
+		hash: map[string]bool{},
+	}
+}
+
+const (
+	ConcurrencyNum = 32
```
[the remark this answers is not in the document]
Comment (riita10069) [reply internal/zombie/checker.go:26]: Don't you mean runtime.NumCPU?

File: internal/zombie/checker.go (hunk 3 above)
Comment (riita10069) [reply internal/zombie/checker.go:78]: So does that mean we use destination as the Key of the Map?

Comment (sanposhiho) [reply internal/zombie/checker.go:78]: > use destination as the Key of the Map?
Yes!

File: internal/zombie/checker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sanposhiho) [reply internal/zombie/checker.go:26]: No

> execute user-level Go code simultaneously. There is no limit to the number of threads that can be blocked in system calls on behalf of Go code; those do not count against the GOMAXPROCS limit. This package's GOMAXPROCS function queries and changes the limit.
https://pkg.go.dev/runtime

And its default value is runtime.NumCPU :)

> It defaults to the value of runtime.NumCPU. 

https://pkg.go.dev/runtime#GOMAXPROCS


For example, upper limit of concurrency for `go test` command is `GOMAXPROCS`.
https://github.com/golang/go/blob/master/src/cmd/go/internal/test/test.go#L256

[the remark this answers is not in the document]
Comment (riita10069) [reply internal/zombie/checker.go:26]: I can't understand what the return value of runtime.GOMAXPROCS(0) means.
I think it is same value as runtime.NumCPU. Because that's the default.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply internal/zombie/checker.go:26]: GOMAXPROCS can be passed from environment variables and user can change it.

And, it means that
> The GOMAXPROCS variable limits the number of operating system threads that can execute user-level Go code simultaneously. There is no limit to the number of threads that can be blocked in system calls on behalf of Go code; those do not count against the GOMAXPROCS limit. This package's GOMAXPROCS function queries and changes the limit.
https://pkg.go.dev/runtime#hdr-Environment_Variables

[the remark this answers is not in the document]
Comment (sanposhiho) [reply internal/zombie/checker.go:26]: runtime.GOMAXPROCS(0) returns the current GOMAXPROCS
> If n < 1, it does not change the current setting
https://pkg.go.dev/runtime#GOMAXPROCS

File: internal/command/check.go (hunk 7 above)
Comment (riita10069) [reply internal/command/check.go:57]: In Source Mode, it is returning immediately and will not move to #L71.

Why can't I use the current implementation?

Comment (riita10069) [reply internal/command/check.go:57]: I decided to give up on the parallelism for now.

**Comment by riita10069** [comment]
Verified that it works locally.
Should we implement automated tests?

File: internal/command/check.go (hunk 7 above)
Comment (sanposhiho) [reply internal/command/check.go:52]: two reasons why I think it is better to move it to L84.

- we don't want to write the same logic two times. `zombieChecker.Search()` is needed for both sourcemode and reflectmode and it will be executed L84 even if we don't execute it here.
- we don't want to run `zombieChecker.Search()` if destination don't match with `target_dir`

Yes, as you said, we can find un-generated mocks here. But we can find it even if we move it to L84, and the time cost remains the same in both cases. 

Are there any other reasons you think it is better to execute `zombieChecker.Search()` here?

File: internal/command/check.go [hunk 9]
```diff
@@ -70,6 +81,12 @@ func (r Runner) Check() {
 				}
 			}
 
+			key := runner.GetDestination()
+			if ok := zombieChecker.Search(key); !ok {
+				fmt.Fprintf(os.Stderr, "[ERROR] mock has not been generated. destination: %s\n", key)
+				isFail = true
```
Comment (sanposhiho) [inline internal/command/check.go:87]: need `return nil` after this maybe?

**Review by sanposhiho — COMMENTED** [review]
thanks for fixing some places

File: internal/command/check.go (hunk 7 above)
Comment (riita10069) [reply internal/command/check.go:52]: Because in Source Mode if it returns at L61.., L84 is not executed.

**Comment by sanposhiho** [comment]
> Should we implement automated tests?

Of course, if have tests, it is great. :)

But, I don't write any tests so far, because of feeling めんどくさい... 
So if you also feel that, I say nothing about test on this PR. And I will write tests someday....

**Comment by riita10069** [comment]
I empathize with you so much about めんどくさい lol.
Because A virtual file system may need to be implemented for e2e testing.
So I would like to think about e2e testing as the other PR.

File: internal/command/check.go (hunk 7 above)
Comment (sanposhiho) [reply internal/command/check.go:52]: Aha, OK. Now I understood your thought. 
What you want to say is we should check un-generated mock before source checksum check, right? That's a fair point. 👍 

It is ok that sourceChecksum check is on here because it does only skip. But with the current code, we print an un-generated error for mocks on non-target dir. It's what we don't want to.

So let's move sourceChecksum check after L89 and do this check only when `m.Mode == model.SourceMode` so that we can delete L51-55.

Comment (riita10069) [reply internal/command/check.go]: OK, I'll do so

File: internal/zombie/checker.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/sanposhiho/gomockhandler/internal/util"
+)
+
+type Checker struct {
+	hash map[string]bool
```
Comment (sanposhiho) [inline internal/zombie/checker.go:16]: IMO we can use the more appropriate name on this `hash` field (because it doesn't have a hash.)

`isManagedMock` or `managedMockMap` is a more understandable name, maybe? 
(I don't have the best idea. If you can think of a better name, please use it :))

File: internal/command/check.go [hunk 11]
```diff
@@ -30,8 +31,13 @@ func (r Runner) Check() {
 		log.Fatalf("failed to change dir: %v", err)
 	}
 
+	zombieChecker := zombie.NewChecker()
+	if err = zombieChecker.FindMocks(ctx); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to find mock list generated by gomock: %v\n", err)
```
[the remark this answers is not in the document]
Comment (riita10069) [reply internal/command/check.go:36]: IMO:
Exit(1) should not be executed by some goroutine.
Exit of the whole process should be limited to a specific line of the main function.
However, it is possible for any goroutine to write to StdOut, StdErr.
The reason is to make it easier to control the termination of any goroutine properly.

You don't think so?

[the remark this answers is not in the document]
Comment (riita10069) [reply internal/command/check.go:36]: I noticed that the Check command uses log.Fatal for error handling, so it is better to follow it. so, I'll do so.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply internal/command/check.go]: > I noticed that the Check command uses log.Fatal for error handling, so it is better to follow it. so, I'll do so.

---

(off topic, you don't need to reply to it)

> Exit(1) should not be executed by some goroutine.

This is definitely yes. In many case we should not call Exit() on goroutine.

> Exit of the whole process should be limited to a specific line of the main function.

(On huge programs, it is better to do so, but) This is partially No (in my knowledges. Plz correct me if I wrong or I miss something 🙏 )

Most of the reasons why we should not use Exit() is because defer funcs will not be executed when Exit(). 
In other words, we can use Exit() if we don't use defer funcs on the caller funcs.

**Review by sanposhiho — APPROVED** [review]
Looks good 👍 
Thanks for your big contribution!!

**Comment by sanposhiho** [comment]
@riita10069 I changed the branch protection rule so that this PR can be merged. 
Please merge it If you can.

**Comment by riita10069** [comment]
Thanks for your review.
I don't have repository privileges, so I can't merge.
Only those with write access to this repository can merge pull requests.

**Comment by sanposhiho** [comment]
> Only those with write access to this repository can merge pull requests.

Aha, Okay. (I forgot about an access right)
I'll merge it.
